### ISSUE TITLE
`near_bindgen` methods with no `self` check for attached deposit

### ISSUE THREAD

**Issue by austinabell** [near/near-sdk-rs#718]
update: as per https://github.com/near/near-sdk-rs/pull/774#issuecomment-1561817582, what should be done to resolve this issue, 
has changed. An implementor should check if https://github.com/near/near-sdk-rs/pull/774#issuecomment-1561817582 still holds true , i.e. if calling `near_sdk::env::attached_deposit()`  from view methods still doesn't panic.

---

Methods like this will add the codegen assumed when a contract mutates state. A slightly strange pattern we have that defaults to checking this only when the parameter is `&mut self` even though in any function there can be state changes not tied to the contract state. 

Tying the check for attached deposit to whether or not contract state changes is a strange pattern IMO but the band-aid fix is to just remove this check when contract state isn't used. This needs to be done so that these methods can be called with a view operation, otherwise, the runtime will fail when trying to call `env::attached_deposit` from within a view call.

```rust
#[near_bindgen]
pub struct SomeContract;

#[near_bindgen]
impl SomeContract {
    pub fn get_val(s: &String) -> Option<u8> 

    { // ... } 

// ...
```

Test for this looks like:

```rust
#[test]
fn no_self_no_args() {
    let impl_type: Type = syn::parse_str("Hello").unwrap();
    let mut method: ImplItemMethod = syn::parse_str("pub fn method() { }").unwrap();
    let method_info = ImplItemMethodInfo::new(&mut method, impl_type).unwrap();
    let actual = method_info.method_wrapper();
    let expected = quote!(
       #[cfg(target_arch = "wasm32")]
       #[no_mangle]
       pub extern "C" fn method() 

       { near_sdk::env::setup_panic_hook(); Hello::method(); } 

     );
    assert_eq!(expected.to_string(), actual.to_string());
}
```

which is currently failing because it adds the extra check

**Comment by itegulov** [comment]
Just to be clear, codegen having a check for the attached deposit is a side effect of such functions being identified as call functions.

<sub>Also, the markdown is all messed up again, curse you exalate</sub>

**Comment by HumbertoTM10** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/near-sdk-rs).

### My background and how it can be leveraged
I have 4 years working in development environments, have encountered several challenges, requests and teams focusing in reusable and efficient code, so that gives me a better understanding of the code and how to solve problems efficiently.


### How I plan on tackling this issue
 My approach here would be adding checks for env::attached_deposit in NEAR contract methods, the solution involves refining the code generation logic to more accurately distinguish between view methods and state-changing methods.

**Comment by PrincesoDan** [comment]
I am applying to this issue via [OnlyDust platform](https://app.onlydust.com/p/near-sdk-rs).

### My background and how it can be leveraged
I'm developer working around blockchain

**Comment by Krishn1412** [comment]
I’d like to help with this.

**Comment by PrathamJha1** [comment]
Would love to contribute here , given my experience as a JavaScript and Typescript developer working in the Web3 space for over a year.

**Comment by PrincesoDan** [comment]
it's open (?)

**Comment by dj8yfo** [comment]
@PrincesoDan @PrathamJha1 i've updated issue description. 
We'll do a triple review with @akorchyn and @PolyProgrammist just to make sure nothing significant isn't broken, 
if you submit a pr to resolve this issue.

**Comment by dj8yfo** [comment]
@PrathamJha1 copy-pasting link to https://github.com/near/near-sdk-rs/issues/1219#issuecomment-2579748415

**Comment by krishnabandewar** [comment]
Could I take a shot at this?

**Comment by dj8yfo** [comment]
@krishnabandewar yes, sure, this looks more useful than the other issue you commented on.
